blueocean-pipeline-editor/src/main/js/components/editor/EditorMain.jsx

```diff
@@ -170,58 +224,77 @@ export class EditorMain extends Component<DefaultProps, Props, State> {
-            title={<h4>
-                    Pipeline Settings
-                </h4>}>
-            <AgentConfiguration key={'agent'+pipelineStore.pipeline.id} node={pipelineStore.pipeline} onChange={agent => (selectedStage && agent.type == 'none' ? delete pipelineStore.pipeline.agent : pipelineStore.pipeline.agent = agent) && this.pipelineUpdated()} />
-            <EnvironmentConfiguration key={'env'+pipelineStore.pipeline.id} node={pipelineStore.pipeline} onChange={e => this.pipelineUpdated()} />
-        </ConfigPanel>);
-
-        if (globalConfigPanel) sheets.push(globalConfigPanel);
-
-        const stageConfigPanel = selectedStage && (<ConfigPanel className="editor-config-panel stage" key={'stageConfig'+selectedStage.id}
-            onClose={e => pipelineValidator.validate() || this.graphSelectedStageChanged(null)}
-            title={
-                <div>
-                    <input className="stage-name-edit" placeholder="Name your stage" defaultValue={title} 
-                        onChange={e => (selectedStage.name = e.target.value) && this.pipelineUpdated()} />
-                    <MoreMenu>
-                        <a onClick={e => this.deleteStageClicked(e)}>Delete</a>
-                    </MoreMenu>
-                </div>
-            }>
-            <ValidationMessageList node={selectedStage} />
-            <EditorStepList steps={steps}
-                        onAddStepClick={() => this.openSelectStepDialog()}
-                        onAddChildStepClick={parent => this.openSelectStepDialog(parent)}
-                        onStepSelected={(step) => this.selectedStepChanged(step)} />
-            {/*
-            <AgentConfiguration key={'agent'+configurationStage.id} node={configurationStage} onChange={agent => (selectedStage && agent.type == 'none' ? delete configurationStage.agent : configurationStage.agent = agent) && this.pipelineUpdated()} />
-            <EnvironmentConfiguration key={'env'+configurationStage.id} node={configurationStage} onChange={e => this.pipelineUpdated()} />
-            */}
-        </ConfigPanel>);
-
-        if (stageConfigPanel) sheets.push(stageConfigPanel);
+        // Stage config panel
+        if (selectedStage) {
+            // Determine if we need to show a particular configuration page
+            // and what errors to display
+            const sectionErrors = {};
+            sectionErrors.stage = _getStageErrors(selectedStage, 'children', 'steps', 'agent', 'environment', 'when', 'post');
+            sectionErrors.steps = _getStageErrors(selectedStage.steps);
+            sectionErrors.settings = _getStageErrors(selectedStage.agent) || _getStageErrors(selectedStage.environment);
+            sectionErrors.show = sectionErrors.steps ? 'steps' : (sectionErrors.settings ? 'settings' : null);
+
+            sheets.push(
+                <ConfigPanel className="editor-config-panel stage" key={'stageConfig' + selectedStage.id}
```

Comment:
Is the use of `key` here to force remount of the component in the event the selected stage is changed? If so, it might be good to add a comment why that is.

Reply:
👍